docs/backend/widgets.md

```diff
@@ -12,15 +12,652 @@ myst:
+`plone.app.z3cform` renders [a frame around each widget](https://github.com/plone/plone.app.z3cform/blob/master/plone/app/z3cform/templates/widget.pt), which usually consists of:
+
+-   Label
+-   Required marker
+-   Description
+
+You might want to customize this widget frame for your own form.
+Below is an example of how to do it.
+
+Copy [`widget.pt`](https://github.com/plone/plone.app.z3cform/blob/master/plone/app/z3cform/templates/widget.pt) to your own package, and edit it.
+
+Then add the following code to `configure.zcml`.
+
+```xml
+    <browser:page
+        name="ploneform-render-widget"
+        for=".demo.IDemoWidget"
+        class="plone.app.z3cform.templates.RenderWidget"
+        permission="zope.Public"
+        template="demo-widget.pt"
+        />
+
+```
+
+Then create a new marker interface in Python code.
+
+```python
+from zope.interface import Interface
+
+
+class IDemoWidget(Interface):
+    pass
+```
+
+Then apply this marker interface to your widgets in `form.update()`.
+
+```python
+from zope.interface import alsoProvides
+
+class MyForm(...):
+
+    def update(self):
+        super().update()
+        # applies custom widget frame to all widgets in this form instance
+        for widget in form.widgets.values():
+            alsoProvides(widget, IDemoWidget)
+```
+
+
+## Combined widgets
+
+You can combine several widgets into one with `z3c.form.browser.multil.MultiWidget` and `z3c.form.browser.object.ObjectWidget` classes.
+
+The following example shows how to create an input widget with minimum and maximum values.
+
+```python
+import zope.interface
+import zope.schema
+from zope.schema.fieldproperty import FieldProperty
+
+import z3c.form
+from z3c.form.object import registerFactoryAdapter
+
+
+class IMinMax(zope.interface.Interface):
+    """ Helper schema for min and max fields """
+
+    min = zope.schema.Float(required=False)
+
+    max = zope.schema.Float(required=False)
+
+
+@zope.interface.implementer(IMinMax)
+class MinMax(object):
+    """ Store min-max field values """
+    min = FieldProperty(IMinMax["min"])
+    max = FieldProperty(IMinMax["max"])
+
+
+registerFactoryAdapter(IMinMax, MinMax)
+
+
+field = zope.schema.Object(
+    __name__="minmax",
+    title=label,
+    schema=IMinMax,
+    required=False
+)
+```
+
+Then you mark the widget in `updateWidgets()`:
+
+```python
+def updateWidgets(self):
+    """
+    """
+
+    super(FilteringGroup, self).updateWidgets()
+
+    # Add min and max CSS class rendering hints
+    for widget in self.widgets.values():
+        if isinstance(widget, z3c.form.browser.object.ObjectWidget):
+            widget.template = Z3ViewPageTemplateFile("templates/minmax.pt")
+            widget.addClass("min-max-widget")
+            zope.interface.alsoProvides(widget, IFilterWidget)
+```
+
+And then the page template which renders both 0. widget (min) and 1. widget (max) on the same line.
+
+``` {code-block} html
```

Comment:
The only reason to use `{codeblock}` is when you want to use Sphinx attributes, such as `emphasize-lines`.

````suggestion
```html
````